File: android/electrode-reactnative-bridge/src/main/java/com/walmartlabs/electrode/reactnative/bridge/ElectrodeBridgeTransceiver.java
package com.walmartlabs.electrode.reactnative.bridge;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.VisibleForTesting;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.walmartlabs.electrode.reactnative.bridge.helpers.Logger;

import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Class that is responsible for transmitting messages between native and react native
 */
class ElectrodeBridgeTransceiver extends ReactContextBaseJavaModule implements ElectrodeBridge, ElectrodeReactBridge {

    private static final String TAG = ElectrodeBridgeTransceiver.class.getSimpleName();

    private final ReactContextWrapper mReactContextWrapper;
    private final EventDispatcher mEventDispatcher;
    private final RequestDispatcher mRequestDispatcher;

    // Singleton instance of the bridge
    private static ElectrodeBridgeTransceiver sInstance;

    private final ConcurrentHashMap<String, BridgeTransaction> pendingTransactions = new ConcurrentHashMap<>();
    private final EventRegistrar<ElectrodeBridgeEventListener<Bundle>> mEventRegistrar = new EventRegistrarImpl<>();
    private final RequestRegistrar<ElectrodeBridgeRequestHandler<Bundle, Object>> mRequestRegistrar = new RequestRegistrarImpl<>();

    private static boolean sIsReactNativeReady;

    /**
     * Initializes a new instance of ElectrodeBridgeTransceiver
     *
     * @param reactContextWrapper The react application context
     */
    private ElectrodeBridgeTransceiver(@NonNull ReactContextWrapper reactContextWrapper) {
        super(reactContextWrapper.getContext());
        mReactContextWrapper = reactContextWrapper;
        mEventDispatcher = new EventDispatcherImpl(mEventRegistrar);
        mRequestDispatcher = new RequestDispatcherImpl(mRequestRegistrar);
    }

    /**
     * Creates the ElectrodeBridgeTransceiver singleton
     *
     * @param reactApplicationContext The react application context
     * @return The singleton instance of ElectrodeBridgeTransceiver
     */
    static ElectrodeBridgeTransceiver create(ReactApplicationContext reactApplicationContext) {
        return create(new ReactContextWrapperInternal(reactApplicationContext));
    }

    /**
     * Creates the ElectrodeBridgeTransceiver singleton
     *
     * @param reactContextWrapper {@link ReactContextWrapper}
     * @return The singleton instance of ElectrodeBridgeTransceiver
     */
    @VisibleForTesting
    static ElectrodeBridgeTransceiver create(@NonNull ReactContextWrapper reactContextWrapper) {
        Logger.d(TAG, "Creating ElectrodeBridgeTransceiver instance");
        synchronized (ElectrodeBridgeTransceiver.class) {
            if (sInstance == null) {
                sInstance = new ElectrodeBridgeTransceiver(reactContextWrapper);
            }
        }
        return sInstance;
    }

    /**
     * Returns the singleton instance of the bridge
     */
    public static ElectrodeBridgeTransceiver instance() {
        if (sInstance == null) {
            throw new IllegalStateException("Bridge transceiver instance has not been created yet. Transceiver requires a valid ReactContext to get initialized. Make sure to that BridgePackage is added to react module which normally initializes transceiver.");
        }
        return sInstance;
    }

    /**
     * @return the name of this module. This will be the name used to {@code require()} this module
     * from javascript.
     */
    @Override
    public String getName() {
        return "ElectrodeBridge";
    }

    @NonNull
    @Override
    public UUID addEventListener(@NonNull String name, @NonNull ElectrodeBridgeEventListener<Bundle> eventListener) {
        Logger.d(TAG, "Adding eventListener(%s) for event(%s)", eventListener, name);
        return mEventRegistrar.registerEventListener(name, eventListener);
    }

    @Override
    public void registerRequestHandler(@NonNull String name, @NonNull ElectrodeBridgeRequestHandler<Bundle, Object> requestHandler) {
        mRequestRegistrar.registerRequestHandler(name, requestHandler);
    }

    /**
     * Emits an event with some data to the JS react native side
     *
     * @param event The event to emit
     */
    @SuppressWarnings("unused")
    @Override
    public void sendEvent(@NonNull ElectrodeBridgeEvent event) {
        Logger.d(TAG, "Emitting event[name:%s id:%s]", event.getName(), event.getId());

        notifyReactEventListeners(event);
        notifyLocalEventListeners(event);
    }


    /**
     * Sends a request
     *
     * @param request          The request to send
     * @param responseListener Listener to be called upon request completion
     */
    @SuppressWarnings("unused")
    @Override
    public void sendRequest(@NonNull final ElectrodeBridgeRequest request, @NonNull final ElectrodeBridgeResponseListener<Bundle> responseListener) {
        handleRequest(request, responseListener);
    }

    /**
     * This method is used by react native to dispatch an event on the native side.
     * <p>
     * This could be a REQUEST, RESPONSE, or an EVENT
     *
     * @param data The event data
     */
    @ReactMethod
    @Override
    public void sendMessage(@NonNull final ReadableMap data) {
        Logger.d(TAG, "received message from JS(data=%s)", data);
        BridgeMessage.Type type = BridgeMessage.Type.getType(data.getString(BridgeMessage.BRIDGE_MSG_TYPE));
        if (type != null) {
            switch (type) {
                case EVENT:
                    ElectrodeBridgeEvent event = ElectrodeBridgeEvent.create(data);
                    if (event != null) {
                        Logger.d(TAG, "Received message is an EVENT(name=%s), will notify local event listeners.", event.getName());
                        notifyLocalEventListeners(event);
                    } else {
                        throw new IllegalArgumentException("Unable to construct event from data");
                    }
                    break;
                case REQUEST:
                    ElectrodeBridgeRequest request = ElectrodeBridgeRequest.create(data);
                    if (request != null) {
                        Logger.d(TAG, "Received message is a REQUEST(name=%s), will look for a request handler and forward this request", request.getName());
                        handleRequest(request, null);
                    } else {
                        throw new IllegalArgumentException("Unable to construct request from data");
                    }

                    break;
                case RESPONSE:
                    ElectrodeBridgeResponse response = ElectrodeBridgeResponse.create(data);
                    if (response != null) {
                        Logger.d(TAG, "Received message is a RESPONSE for a request(name=%s, id=%s)", response.getName(), response.getId());
                        handleResponse(response);
                    } else {
                        throw new IllegalArgumentException("Unable to construct a response from data");
                    }

                    break;
            }
        } else {
            throw new IllegalArgumentException("Unable to identify request type. Should never reach here.");
        }
    }

    private void handleRequest(@NonNull final ElectrodeBridgeRequest request, @Nullable ElectrodeBridgeResponseListener<Bundle> responseListener) {
        logRequest(request);

        if (responseListener == null && !request.isJsInitiated()) {
            throw new IllegalArgumentException("A response lister is required for a non-JS initiated request");
        }

        final BridgeTransaction bridgeTransaction = createTransaction(request, responseListener);

        if (mRequestDispatcher.canHandleRequest(request.getName())) {
            dispatchRequestToLocalHandler(bridgeTransaction);
        } else if (!request.isJsInitiated()) {//GOTCHA: Should not send a request back JS if it was initiated from JS side.
            dispatchRequestToReact(bridgeTransaction);
        } else {
            Logger.d(TAG, "No handler available to handle the request(id=%s, name=%s). Will fail the request", request.getId(), request.getName());
            handleResponse(ElectrodeBridgeResponse.createResponseForRequest(request, null, BridgeFailureMessage.create("ENOHANDLER", "No registered request handler found for " + request.getName())));
        }
    }

    @NonNull
    private BridgeTransaction createTransaction(@NonNull ElectrodeBridgeRequest request, @Nullable ElectrodeBridgeResponseListener<Bundle> responseListener) {
        final BridgeTransaction bridgeTransaction = new BridgeTransaction(request, responseListener);
        pendingTransactions.put(request.getId(), bridgeTransaction);
        startTimeOutCheckForTransaction(bridgeTransaction);
        return bridgeTransaction;
    }

    private void startTimeOutCheckForTransaction(@NonNull final BridgeTransaction transaction) {
        Handler handler = new Handler(Looper.getMainLooper());
        handler.postDelayed(new Runnable() {
            public void run() {
                Logger.d(TAG, "Checking timeout for request(id=%s)", transaction.getRequest().getId());
                handleResponse(ElectrodeBridgeResponse.createResponseForRequest(transaction.getRequest(), null, BridgeFailureMessage.create("EREQUESTTIMEOUT", "Request timeout")));
            }
        }, transaction.getRequest().getTimeoutMs());
    }

    private void dispatchRequestToLocalHandler(@NonNull final BridgeTransaction transaction) {
        Logger.d(TAG, "Sending request(id=%s) to local handler", transaction.getRequest().getId());

        final ElectrodeBridgeRequest request = transaction.getRequest();
        mRequestDispatcher.dispatchRequest(transaction.getRequest(), new ElectrodeBridgeResponseListener<Object>() {
            @Override
            public void onFailure(@NonNull FailureMessage failureMessage) {
                ElectrodeBridgeResponse response = ElectrodeBridgeResponse.createResponseForRequest(request, null, failureMessage);
                handleResponse(response);
            }

            @Override
            public void onSuccess(@Nullable Object responseData) {
                ElectrodeBridgeResponse response = ElectrodeBridgeResponse.createResponseForRequest(request, responseData, null);
                handleResponse(response);
            }
        });

    }

    private void dispatchRequestToReact(@NonNull BridgeTransaction bridgeTransaction) {
        Logger.d(TAG, "Sending request(id=%s) over to JS side as there is no local request handler available", bridgeTransaction.getId());
        mReactContextWrapper.emitEvent(bridgeTransaction.getRequest());
    }

    private void handleResponse(@NonNull ElectrodeBridgeResponse bridgeResponse) {
        Logger.d(TAG, "Handling bridge response");
        BridgeTransaction transaction = pendingTransactions.get(bridgeResponse.getId());
        if (transaction != null) {
            transaction.setResponse(bridgeResponse);
            completeTransaction(transaction);
        } else {
            Logger.i(TAG, "Response(id=%s, name=%s) will be ignored as the transaction for this request has already been removed from the queue. Perhaps it's already timed-out or completed", bridgeResponse.getId(), bridgeResponse.getName());
        }

    }

    private void notifyLocalEventListeners(@NonNull final ElectrodeBridgeEvent event) {
        mReactContextWrapper.runOnUiQueueThread(new Runnable() {
            @Override
            public void run() {
                mEventDispatcher.dispatchEvent(event);
            }
        });
    }

    private void notifyReactEventListeners(@NonNull ElectrodeBridgeEvent event) {
        mReactContextWrapper.emitEvent(event);
    }

    private void completeTransaction(@NonNull final BridgeTransaction transaction) {
        if (transaction.getResponse() == null) {
            throw new IllegalArgumentException("Cannot complete transaction, a transaction can only be completed with a valid response.");
        }
        Logger.d(TAG, "completing transaction(%s)", transaction.getId());

        pendingTransactions.remove(transaction.getId());

        final ElectrodeBridgeResponse response = transaction.getResponse();
        logResponse(response);

        if (transaction.isJsInitiated()) {
            Logger.d(TAG, "Completing by emitting event to JS since the request was initiated from JS side.");
            mReactContextWrapper.emitEvent(response);
        } else {
            if (transaction.getFinalResponseListener() != null) {
                if (response.getFailureMessage() != null) {
                    Logger.d(TAG, "Completing by issuing a failure call back to local response listener.");
                    mReactContextWrapper.runOnUiQueueThread(new Runnable() {
                        @Override
                        public void run() {
                            transaction.getFinalResponseListener().onFailure(response.getFailureMessage());
                        }
                    });
                } else {
                    Logger.d(TAG, "Completing by issuing a success call back to local response listener.");
                    mReactContextWrapper.runOnUiQueueThread(new Runnable() {
                        @Override
                        public void run() {
                            transaction.getFinalResponseListener().onSuccess(response.bundle());
                        }
                    });
                }
            } else {
                throw new IllegalArgumentException("Should never reach here, a response listener should always be set for a local transaction");
            }

        }

    }

    private void logRequest(@NonNull ElectrodeBridgeRequest request) {
        Logger.d(TAG, "--> --> --> --> --> Request(id=%s, name=%s, isJS=%s)", request.getId(), request.getName(), request.isJsInitiated());
    }

    private void logResponse(ElectrodeBridgeResponse response) {
        Logger.d(TAG, "<-- <-- <-- <-- <-- Response(id=%s, name=%s, data=%s, error=%s) received", response.getId(), response.getName(), response.getData(), response.getFailureMessage());
    }

    public interface ReactNativeReadyListener {
        void onReactNativeReady();
    }

    private static ReactNativeReadyListener sReactNativeReadyListener;

    public static void registerReactNativeReadyListener(ReactNativeReadyListener listener) {
        // If react native initialization is already completed, just call listener
        // immediately
        if (sIsReactNativeReady) {
            listener.onReactNativeReady();
        }
        // Else it will get invoked whenever react native initialization is done
        else {
            sReactNativeReadyListener = listener;
        }
    }

    public void onReactNativeInitialized() {
        sIsReactNativeReady = true;
        if (sReactNativeReadyListener != null) {
            sReactNativeReadyListener.onReactNativeReady();
        }
    }
}

File: ios/ElectrodeReactNativeBridge/ElectrodeBridgeTransceiver.m
//
//  ElectrodeBridgeTransceiver.m
//  ElectrodeReactNativeBridge
//
//  Created by Claire Weijie Li on 3/22/17.
//  Copyright © 2017 Walmart. All rights reserved.
//

#import "ElectrodeBridgeTransceiver.h"
#import "ElectrodeBridgeTransceiver_Internal.h"
#import "ElectrodeEventDispatcherNew.h"
#import "ElectrodeRequestDispatcherNew.h"
#import "ElectrodeBridgeTransaction.h"
#import "ElectrodeEventRegistrarNew.h"
#import "ElectrodeRequestRegistrarNew.h"
#import <React/RCTLog.h>
#import <React/RCTBridge.h>
#import "ElectrodeBridgeMessage.h"
#import "ElectrodeBridgeEventNew.h"

NS_ASSUME_NONNULL_BEGIN

@interface ElectrodeBridgeResponseListenerImplementor : NSObject<ElectrodeBridgeResponseListener>

@property(nonatomic, copy) ElectrodeBridgeResponseListenerSuccessBlock successBlock;
@property(nonatomic, copy) ElectrodeBridgeResponseListenerFailureBlock failureBlock;

-(instancetype)initWithSuccessBlock: (ElectrodeBridgeResponseListenerSuccessBlock) successBlock
                       failureBlock: (ElectrodeBridgeResponseListenerFailureBlock) failureBlock;
@end

@implementation ElectrodeBridgeResponseListenerImplementor

-(instancetype)initWithSuccessBlock: (ElectrodeBridgeResponseListenerSuccessBlock) successBlock
                       failureBlock: (ElectrodeBridgeResponseListenerFailureBlock) failureBlock {
    if (self = [super init]) {
        _successBlock = successBlock;
        _failureBlock = failureBlock;
    }
    return self;
}

-(void)onFailure:(id<ElectrodeFailureMessage>)failureMessage
{
    if (self.failureBlock) {
        self.failureBlock(failureMessage);
    }
}

-(void)onSuccess:(NSDictionary * _Nullable)responseData
{
    if(self.successBlock) {
        self.successBlock(responseData);
    }
}

@end

@interface ElectrodeBridgeTransceiver()

@property(nonatomic, copy) NSString *name;
@property(nonatomic, strong) ElectrodeEventDispatcherNew *eventDispatcher;
@property(nonatomic, strong) ElectrodeRequestDispatcherNew *requestDispatcher;
@property(nonatomic, copy) NSMutableDictionary<NSString *, ElectrodeBridgeTransaction * > *pendingTransaction;
@property (nonatomic, assign) dispatch_queue_t syncQueue; //this is used to make sure access to pendingTransaction is thread safe.


@end
@implementation ElectrodeBridgeTransceiver

+(instancetype)sharedInstance {
    return sharedInstance;
}

-(instancetype)init {
    if (self = [super init])
    {
        ElectrodeRequestRegistrarNew *requestRegistrar = [[ElectrodeRequestRegistrarNew alloc] init];
        ElectrodeEventRegistrarNew *eventRegistrar = [[ElectrodeEventRegistrarNew alloc] init];
        _requestDispatcher = [[ElectrodeRequestDispatcherNew alloc] initWithRequestRegistrar:requestRegistrar];
        _eventDispatcher = [[ElectrodeEventDispatcherNew alloc] initWithEventRegistrar:eventRegistrar];
        _pendingTransaction = [[NSMutableDictionary alloc] init];
    }
    return self;
}
RCT_EXPORT_MODULE(ElectrodeBridge);
+ (NSArray *)electrodeModules
{
    return @[[[ElectrodeBridgeTransceiver alloc] init]];
}

-(NSArray *) supportedEvents
{
    return @[@"electrode.bridge.message"];
}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////
#pragma ElectrodeNativeBridge implementation
-(void)sendRequest:(ElectrodeBridgeRequestNew *)request withResponseListener:(id<ElectrodeBridgeResponseListener> _Nonnull) responseListener {
    [self handleRequest:request withResponseListner:responseListener];
}

-(NSUUID *)regiesterRequestHandlerWithName: (NSString *)name handler:(id<ElectrodeBridgeRequestHandler>) requestHandler error:(NSError * _Nullable __autoreleasing * _Nullable)error{
    NSUUID *uUID = [self.requestDispatcher.requestRegistrar registerRequestHandler:name requestHandler:requestHandler error:error]; //CLAIRE TODO: Check if need to return UUID
    return uUID;
}

-(void)resetRegistrar {
    [self.requestDispatcher.requestRegistrar reset];
}

-(void)sendEvent: (ElectrodeBridgeEventNew *)event {
    NSLog(@"ElectrodeBridgeTransceiver: emit event named: %@, id: %@", event.name, event.messageId);
    [self notifyNativeEventListenerWithEvent:event];
    [self notifyReactNativeEventListenerWithEvent:event];
}

-(NSUUID *)addEventListenerWithName: (NSString *)name eventListener: (id<ElectrodeBridgeEventListener>) eventListener {
    NSLog(@"%@, Adding eventListener %@ for event %@", NSStringFromClass([self class]), eventListener, name);
    NSUUID *uUID = [self.eventDispatcher.eventRegistrar registerEventListener:name eventListener:eventListener];
    return uUID;
}
#pragma ElectrodeReactBridge

RCT_EXPORT_METHOD(sendMessage:(NSDictionary *)bridgeMessage)
{
    NSLog(@"Received message from JS(data=%@)", bridgeMessage);
    NSString *typeString = (NSString *)[bridgeMessage objectForKey:kElectrodeBridgeMessageType];
    ElectrodeMessageType type = [ElectrodeBridgeMessage typeFromString:typeString];
    switch (type) {
        case ElectrodeMessageTypeRequest:
        {
            ElectrodeBridgeRequestNew *request = [ElectrodeBridgeRequestNew createRequestWithData:bridgeMessage];
            [self handleRequest:request withResponseListner:nil];
            break;
        }
            
        case ElectrodeMessageTypeResponse:
        {
            ElectrodeBridgeResponse *response = [ElectrodeBridgeResponse createResponseWithData:bridgeMessage];
            if (response != nil) {
                [self handleResponse:response];
            } else {
                [NSException raise:@"invalue resonse" format:@"cannot construct a response from data"];
            }
            break;
        }
        case ElectrodeMessageTypeEvent:
        {
            ElectrodeBridgeEventNew *event = [ElectrodeBridgeEventNew createEventWithData:bridgeMessage];
            if (event != nil) {
                [self notifyNativeEventListenerWithEvent:event];
            } else {
                [NSException raise:@"invalue event" format:@"cannot construct an event from data"];
                
            }
            break;
        }
        case ElectrodeMessageTypeUnknown:
        default:
            [NSException raise:@"invalue message" format:@"cannot construct any message from data"];
            break;
    }
}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////
#pragma private methods

-(void)emitMessage:(ElectrodeBridgeMessage * _Nonnull)bridgeMessage
{
    NSLog(@"Sending bridgeMessage(%@) to JS", bridgeMessage);
    [self sendEventWithName:@"electrode.bridge.message" body:[bridgeMessage toDictionary]];
}

-(void)notifyReactNativeEventListenerWithEvent: (ElectrodeBridgeEventNew *) event {
    [self emitMessage:event];
}

-(void)notifyNativeEventListenerWithEvent: (ElectrodeBridgeEventNew *)event {
    [self.eventDispatcher dispatchEvent:event];
}

-(void)handleRequest:(ElectrodeBridgeRequestNew *)request withResponseListner:(id<ElectrodeBridgeResponseListener> _Nullable) responseListener
{
    [self logRequest:request];
    
    if (responseListener != nil && ![responseListener conformsToProtocol:@protocol(ElectrodeBridgeResponseListener)]) {
        [NSException raise:@"invalid response listener" format:@"A response lister need to conform to ElectrodeBridgeResponseListener protocol"];
    }
    
    if (responseListener == nil && !request.isJsInitiated) {
        [NSException raise:@"invalid operation" format:@"A response lister is required for a native initiated request"];
    }
    ElectrodeBridgeTransaction *transaction = [self createTransactionWithRequest:request responseListner:responseListener];
    if ([self.requestDispatcher canHandlerRequestWithName:request.name] ) {
        [self dispatchRequestToNativeHandlerForTransaction:transaction];
    } else if (!request.isJsInitiated) { //GOTCHA: Make sure not send a request back to JS if it's initiated on JS side
        [self dispatchRequestToReactHandlerForTransaction:transaction];
    }else {
        NSLog(@"No handler available to handle request(%@)", request);
        id<ElectrodeFailureMessage> failureMessage = [ElectrodeBridgeFailureMessage createFailureMessageWithCode:@"ENOHANDLER" message:@"No registered request handler found"];
        ElectrodeBridgeResponse *response = [ElectrodeBridgeResponse createResponseForRequest:request withResponseData:nil withFailureMessage:failureMessage];
        [self handleResponse:response];
    }
}

-(ElectrodeBridgeTransaction *)createTransactionWithRequest: (ElectrodeBridgeRequestNew *)request
                    responseListner:(id<ElectrodeBridgeResponseListener> _Nullable) responseListener
{
    ElectrodeBridgeTransaction *transaction = [[ElectrodeBridgeTransaction alloc] initWithRequest:request responseListener:responseListener];
    
    @synchronized (self) {
        [self.pendingTransaction setObject:transaction forKey:request.messageId];
        [self startTimeOutCheckForTransaction:transaction];
    }
    
    return transaction;
}

-(void)startTimeOutCheckForTransaction: (ElectrodeBridgeTransaction *)transaction
{
    // Add the timeout handler
    __weak ElectrodeBridgeTransceiver *weakSelf = self;
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(transaction.request.timeoutMs * NSEC_PER_MSEC)), dispatch_get_main_queue(), ^{
        
        id<ElectrodeBridgeResponseListener> responseListener = transaction.finalResponseListener;
        if (responseListener && [responseListener conformsToProtocol:@protocol(ElectrodeBridgeResponseListener)])
        {
            id<ElectrodeFailureMessage> failureMessage = [ElectrodeBridgeFailureMessage createFailureMessageWithCode:@"TIMEOUT" message:@"transaction timed out for request"];
            ElectrodeBridgeResponse *response = [ElectrodeBridgeResponse createResponseForRequest:transaction.request withResponseData:nil withFailureMessage:failureMessage];
            [weakSelf handleResponse:response];
        }
    });
}

-(void)dispatchRequestToNativeHandlerForTransaction: (ElectrodeBridgeTransaction *)transaction
{
    NSLog(@"Sending request(%@) to native handler", transaction.request);
    ElectrodeBridgeRequestNew *request = transaction.request;
    __weak ElectrodeBridgeTransceiver *weakSelf = self;

    id<ElectrodeBridgeResponseListener> responseListener = [[ElectrodeBridgeResponseListenerImplementor alloc] initWithSuccessBlock:^(id _Nullable data) {
        ElectrodeBridgeResponse *response = [ElectrodeBridgeResponse createResponseForRequest:request
                                                                             withResponseData:data
                                                                           withFailureMessage:nil];
        [weakSelf handleResponse:response];
    } failureBlock:^(id<ElectrodeFailureMessage> _Nonnull message) {
        ElectrodeBridgeResponse *response = [ElectrodeBridgeResponse createResponseForRequest:request
                                                                             withResponseData:nil
                                                                           withFailureMessage:message];
        [weakSelf handleResponse:response];
    }];
    [self.requestDispatcher dispatchRequest:request withResponseListener:responseListener];
    
}

-(void)dispatchRequestToReactHandlerForTransaction:(ElectrodeBridgeTransaction *)transaction
{
    NSLog(@"Sending request(%@) over to JS handler because there is no local request handler available", transaction.request);
    [self emitMessage:transaction.request];
}

-(void)handleResponse:(ElectrodeBridgeResponse *)response
{
    NSLog(@"hanlding bridge response: %@", response);
    ElectrodeBridgeTransaction *transaction;
    @synchronized (self) {
         transaction = (ElectrodeBridgeTransaction *) [self.pendingTransaction objectForKey:response.messageId];
    }
    if (transaction != nil) {
        transaction.response = response;
        [self completeTransaction:transaction];
    } else {
        NSLog(@"Response(%@) will be ignored because the transcation for this request has been removed from the queue. Perhaps it's already timed-out or completed.", response);
    }
    
}

-(void)completeTransaction:(ElectrodeBridgeTransaction *)transaction
{
    if(transaction.response == nil) {
        [NSException raise:@"invalid transaction" format:@"Cannot complete transaction, a transaction can only be completed with a valid response"];
    }
    NSLog(@"completing transaction(id=%@", transaction.transactionId);

    [self.pendingTransaction removeObjectForKey:transaction.transactionId];
    
    ElectrodeBridgeResponse *response = transaction.response;
    [self logResponse:response];
    
    if(transaction.isJsInitiated) {
        NSLog(@"Completing transaction by emitting event back to JS since the request is initiated from JS side");
        [self emitMessage:response];
    } else {
        if(transaction.finalResponseListener != nil) {
            if(response.failureMessage != nil) {
                NSLog(@"Completing transaction by issuing a failure call back to local response listener");
                dispatch_async(dispatch_get_main_queue(), ^{
                    [transaction.finalResponseListener onFailure:response.failureMessage];
                });
            } else {
                NSLog(@"Completing transaction by issuing a success call back to local response listener");
                dispatch_async(dispatch_get_main_queue(), ^{
                    [transaction.finalResponseListener onSuccess:response.data];
                });
            }
        } else {
            [NSException raise:@"invalid transaction" format:@"Should never reach here. A response listener should always be set for a local transaction"];
        }
    }
}

-(void)logRequest: (ElectrodeBridgeRequestNew *)request {
    NSLog(@"--> --> --> --> --> Request(%@)", request);
}

-(void)logResponse: (ElectrodeBridgeResponse *)response {
    NSLog(@"<-- <-- <-- <-- <-- Response(%@)", response);
}

+ (void)registerReactNativeReadyListener: (ElectrodeBridgeReactNativeReadyListner) listner
{
    if(isReactNativeReady) {
        if (listner) {
            listner(sharedInstance);
        }
    }
    
    reactNativeReadyListener = [listner copy];
}

- (void)onReactNativeInitialized
{
    isReactNativeReady = YES;
    sharedInstance = self;

    if (reactNativeReadyListener) {
        reactNativeReadyListener(self);
    }
}
+ (BOOL)isReactNativeBridgeReady {
    return isReactNativeReady;
}
@end
NS_ASSUME_NONNULL_END
